homeworks/MarlQ/1-halfedge/src/geometry/halfedge.ts

```diff
@@ -0,0 +1,260 @@
+  // the given mesh.
+  //
+  // context is a persistent status for the given mesh and can be reused
+  // for each of the rendering frames unless the mesh intentionally
+  // calls the resetContext() method.
+  context: Matrix;
+  color: Vector;
+  wireframe: Vector;
+
+  // The following four fields are the key fields to represent half-edge based
+  // meshes.
+  verts: Vertex[]; // a list of vertices
+  edges: Edge[]; // a list of edges
+  faces: Face[]; // a list of faces
+  halfedges: Halfedge[]; // a list of halfedges
+
+  /**
+   * constructor constructs the halfedge-based mesh representation.
+   *
+   * @param {string} data is a text string from an .obj file
+   */
+  constructor(data: string) {
+    // context is initialized as an identity matrix.
+    // prettier-ignore
+    this.context = new Matrix(
+      1, 0, 0, 0,
+      0, 1, 0, 0,
+      0, 0, 1, 0,
+      0, 0, 0, 1,
+    );
+    this.color = new Vector(0, 128, 255, 1);
+    this.wireframe = new Vector(125, 125, 125, 1);
+
+    // load .obj file
+    const indices: number[] = [];
+    const positions: Vector[] = [];
+    const lines = data.split('\n');
+    for (let line of lines) {
+      line = line.trim();
+      const tokens = line.split(' ');
+      switch (tokens[0].trim()) {
+        case 'v':
+          positions.push(
+            new Vector(
+              parseFloat(tokens[1]),
+              parseFloat(tokens[2]),
+              parseFloat(tokens[3]),
+              1
+            )
+          );
+          break;
+        case 'f':
+          // only load indices of vertices
+          for (let i = 1; i < tokens.length; i++) {
+            const vv = tokens[i].split('/');
+            indices.push(parseInt(vv[0]) - 1);
+          }
+          break;
+      }
+    }
+
+    this.verts = []; // an array of Vertex object
+    this.edges = []; // an array of Edge object
+    this.faces = []; // an array of Face object
+    this.halfedges = []; // an array of Halfedge object
+    this.buildMesh(indices, positions);
+  }
+
+  /**
+   * buildMesh builds half-edge based connectivity for the given vertex index buffer
+   * and vertex position buffer.
+   *
+   * @param indices is the vertex index buffer that contains all vertex indices.
+   * @param positions is the vertex buffer that contains all vertex positions.
+   */
+  buildMesh(indices: number[], positions: Vector[]) {
+    // TODO: build the halfedge connectivity.
+    //
+    // We can assume the input mesh is a manifold mesh.
+    
+    // DEBUG
+    // let duplicates = 0;
+    // indices = indices.slice(0, 120); 
+
+    positions.forEach((p, i) => {
+      const vert = new Vertex(p);
+      vert.idx = i;
+      this.verts.push(vert);
+    });
+
+    for(let index = 2; index < indices.length; index+=3) {
+      const face = new Face();
+      face.idx = Math.floor(index/3);
+      this.faces.push(face);
+
+      const halfedges: Halfedge[] = [];
```

Comment:
As discussed in precense, we should already know the size of half edges. Hence preallocating is better than allocating on the fly.